tests/cases/unittests/session.ts

```diff
@@ -0,0 +1,459 @@
+			}
+
+			enqueue(msg: protocol.Request) {
+				this.queue = [msg].concat(this.queue);
+			}
+			
+			handleRequest(msg: protocol.Request) {
+				let response: protocol.Response;
+				try {
+					response = this.executeCommand(msg).response;
+				} catch (e) {
+					this.output(undefined, msg.command, msg.seq, e.toString());
+					return;
+				}
+				if (response) {
+					this.output(response, msg.command, msg.seq);
+				}
+			}
+
+			consumeQueue() {
+				while (this.queue.length > 0) {
+					let elem = this.queue[this.queue.length - 1];
+					this.queue = this.queue.slice(0, this.queue.length - 1);
+					this.handleRequest(elem);
+				}
+			}
+		},
+		InProcClient = class {
+			private server: Session & {enqueue: (msg: protocol.Request) => void};
+			private seq: number = 0;
+			private callbacks: ts.Map<(resp: protocol.Response) => void> = {};
+			private eventHandlers: ts.Map<(args: any) => void> = {};
+
+			handle(msg: protocol.Message): void {
+				if (msg.type === 'response') {
+					let response = <protocol.Response>msg;
+					if (this.callbacks[response.request_seq]) {
+						this.callbacks[response.request_seq](response);
+						delete this.callbacks[response.request_seq];
+					}
+				} else if (msg.type === 'event') {
+					let event = <protocol.Event>msg;
+					this.emit(event.event, event.body);
+				}
+			}
+
+			emit(name: string, args: any): void {
+				if (!this.eventHandlers[name]) {
+					return;
+				}
+				this.eventHandlers[name](args);
+			}
+
+			on(name: string, handler: (args: any) => void): void {
+				this.eventHandlers[name] = handler;
+			}
+
+			connect(session: Session & {enqueue: (msg: protocol.Request) => void}): void {
+				this.server = session;
+			}
+			
+			execute(command: string, args: any, callback: (resp: protocol.Response) => void): void {
+				if (!this.server) {
+					return;
+				}
+				this.seq++;
+				this.server.enqueue({
+					seq: this.seq,
+					type: 'command',
+					command,
+					arguments: args
+				});
+				this.callbacks[this.seq] = callback;
+			}
+		};
+		
+		it('can be constructed and respond to commands', (done) => {
+			let cli = new InProcClient(),
+			session = new InProcSession(cli),
+			toEcho = {
+				data: true
+			},
+			toEvent = {
+				data: false	
+			},
+			responses = 0;
+
+			// Connect the client
+			cli.connect(session);
+			
+			// Add an event handler
+			cli.on('testevent', (eventinfo) => {
+				expect(eventinfo).to.equal(toEvent);
+				responses++;
+				expect(responses).to.equal(1);
+			});
+			
+			// Trigger said event from the server
+			session.event(toEvent, 'testevent');
+			
+			// Queue an echo command
+			cli.execute('echo', toEcho, (resp) => {
+				assert(resp.success, resp.message);
+				responses++;
+				expect(responses).to.equal(2);
+				expect(resp.body).to.deep.equal(toEcho);
+			});
+			
+			// Queue a configure command
+			cli.execute('configure', {
+				hostInfo: 'unit test',
+				formatOptions: {
+					newLineCharacter: '`n'
+				}

```

Comment:
Yep. Trivia: That's the newline character in powershell string literals (go figure). 

The exact value doesn't actually matter for the test (it just checks that the options are sent and responded to), and as far as I know the service isn't supposed to choke on values that aren't `\n` or `\r\n`, so it should be okay.